specs/keys/eth2.md

```diff
@@ -0,0 +1,13 @@
+# Keys and their storage within Ethereum 2.0
+
+## Validator withdrawal and signing keys
+
+A validator's withdrawal and signing keys are elements of the key-tree as described below. They are designed to be stored in [keystore files](./eth2.md) with the idea being that clients need only concern themselves with ingesting a signing-key keystore and that this sufficient for a to launch a validator instance.
```

Comment:
The way I was envisioning handling this problem is as follows:

* There is a single mnemonic underlying both the withdrawal keys and signing keys, knowing this mnemonic provides access to all keys.
* There is a single keystore file which also retains access to all the keys as it contains the master key from the tree. (Users don't necessarily need to generate or store this file if they feel safer just holding onto the mnemonics)
* Each validator instance gets its own keystore file which contains the the signing key for just that validator instance.

The above provides separation into a cold keystore and hot keystore**s**, but comes at the cost of needing one file per validator. This could be seen as it is easier to ensure separation between instances when they use different signing files than when they are issued just an index in the key tree.
* **Pros:**
    * Knowledge of withdrawal key yields signing key
* **Cons:**
    * 1 keystore per validator instance for the signing keys.

## Alternatives

The alternatives, as far as I can tell, are:

* Using the same base mnemonic but appending some (standardised) bytes to each to separate signing and withdrawal key-trees. This would mean that a single mnemonic produces both the signing and withdrawal master nodes in their respective key-trees which could be stored in their own keystores.

    * **Pros:**
        * Single mnemonic
        * Hard separation between signing & withdrawal keys
    * **Cons:**
        * Signing keys cannot be derived from withdrawal keys. The mnemonic becomes the only relation between the two key types.
        * It will be harder to get other projects involved in the BLS standardisation effort to use this derivation standard with such application specific machinery built in.

* Having signing and and withdrawal keytrees be sub-trees in the same greater keytree. This is a family of solutions, the best variant of which (that I can see), has the root of the signing sub-tree be a sibling of the withdrawal keys (with its index offset by a constant). This allows for a single tree-node to provide the all signing and withdrawal keys while offering separable signing keys which don't leak any information about the withdrawal keys.
    * **Pros:**
        * Single withdrawal keystore yields all signing and withdrawal keys
        * Separable hot keystore for signing keys.
        * Uses key-tree as intended - for the separation of keys for different purposes.
    * **Cons:**
        * Knowledge of a withdrawal key does not provide knowledge of the corresponding signing key, knowledge of the parent node is required instead.